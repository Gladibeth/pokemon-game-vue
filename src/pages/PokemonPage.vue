<template>
    <h1 v-if="!pokemon">Por favor espere un momento...</h1>
    <div v-else>
      <h1>Quien es el Pokemon?</h1>
      <!-- Imagen -->
      <PokemonPicture :pokemon-id="pokemon.id" :show-pokemon="showPokemon"/>
      <!-- Opciones -->
      <!--selection-pokemon-->
      <PokemonOptions @selectionPokemon="checkAnswer" :pokemons="pokemonArr"/>

      <template v-if="showAnswer">
        <h2>{{ message }}</h2>
        <button @click="newGame">Nuevo juego</button>
      </template>
    </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions.vue';
import PokemonPicture from '@/components/PokemonPicture.vue';
import getPokemonsOptions from '@/helpers/getPokemonOptions';


export default {
  name: 'PokemonPage',
  components:{
    PokemonPicture,
    PokemonOptions
  },
  data(){
    return{
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: null,
    }
  },
  methods:{
    async mixedPokemons(){
      this.pokemonArr = await getPokemonsOptions();
      //console.log(this.pokemonArr);

      const randomId = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[randomId]
    },
    checkAnswer(selectId){
      this.showAnswer = true;
      this.showPokemon = true;
      if(this.pokemon.id == selectId){
        this.message = `Correcto, este es el Pokemon ${this.pokemon.name}`
      }else{
        this.message = `Oops, el Pokemon era ${this.pokemon.name}`
      }
    },
    newGame(){
      this.pokemonArr = [];
      this.showAnswer = false;
      this.showPokemon = false;
      this.pokemon = null;
      this.mixedPokemons();
    }
  },
  mounted(){
    this.mixedPokemons();
  }
}
</script>

