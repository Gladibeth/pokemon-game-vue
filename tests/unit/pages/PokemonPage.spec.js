import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from "@/pages/PokemonPage"
import { pokemons } from "../mocks/pokemon.mock"
import PokemonPicture from "@/components/PokemonPicture"
import PokemonOptions from "@/components/PokemonOptions"
import getPokemonsOptions from "@/helpers/getPokemonOptions"


describe('PokemonPage component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })

  test('', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Debe llamar a el metod mixedPokemons', () => {
    const mixedPokemonsSpy = jest.spyOn(PokemonPage.methods, 'mixedPokemons')
    wrapper = shallowMount(PokemonPage)
    //console.log(mixedPokemonsSpy)
    expect(mixedPokemonsSpy).toHaveBeenCalled()
  })

  test('debe hacer match con data',()=>{
    // Lo monta en el shapshot sutilmente con stub
    const wrapper = shallowMount(PokemonPage, {
      data() {
       return{
        pokemonArr: pokemons,
        pokemon: pokemons[0],
        showPokemon: false,
        showAnswer: false,
        message: null,
       }
      }
    })

    // lo monta y renderiza los componentes 
    /* const wrapper = mount(PokemonPage, {
      data() {
       return{
        pokemonArr: pokemons,
        pokemon: pokemons[0],
        showPokemon: false,
        showAnswer: false,
        message: null,
       }
      }
    }) */


    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Debe mostrar el los componentes PokemonPicture y PokemonOptions', async () => {


    const wrapper = shallowMount(PokemonPage,{
      data(){
        return{
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: null,
        }
      }
    })
    // Mi resuesta a la tarea
    // PokemonPicture debe existir sin importar el componente hijo
    const pokemonPicture = wrapper.findComponent({name: 'PokemonPicture'});

    // PokemonOptions debe existir Impotando el componente hijo
    const pokemonOptions = wrapper.findComponent(PokemonOptions);

    //console.log(pokemonPicture)
    expect(pokemonPicture.exists()).toBe(true)
    expect(pokemonOptions.exists()).toBe(true)

    //PokemonPicture debe tener el atributo de pokemonId = 1
    expect(pokemonPicture.attributes('pokemonid')).toBe('1')

    // PokemonOptions debe tenet el atributo de pokemons
    expect(pokemonOptions.props().hasOwnProperty('pokemons')).toBe(true)


    // 2 opcion La del profesor
    const picture = wrapper.find('pokemon-picture-stub')
    const options = wrapper.find('pokemon-options-stub')

    expect(picture.exists()).toBe(true)
    expect(options.exists()).toBe(true)

    expect(picture.attributes('pokemonid')).toBe('1')
    expect(options.attributes('pokemons')).toBeTruthy()
  })

  test('Se debe evaluar en checkAnswer', async() => {
    const wrapper = shallowMount(PokemonPage,{
      data(){
        return{
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: null,
        }
      }
    })

    await wrapper.vm.checkAnswer(1)

    expect(wrapper.find('h2').exists()).toBe(true)
    //console.log(wrapper.find('h2').text())

    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.find('h2').text()).toBe(`Correcto, este es el Pokemon ${pokemons[0].name}`)

    //tarea que salga el mensaje negativo cuando sera error
    await wrapper.vm.checkAnswer(5)

    expect(wrapper.vm.message).toBe(`Oops, el Pokemon era ${pokemons[0].name}`)
  })  


})