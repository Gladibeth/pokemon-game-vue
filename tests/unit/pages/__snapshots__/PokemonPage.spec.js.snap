// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component  1`] = `<h1>Por favor espere un momento...</h1>`;

exports[`PokemonPage component debe hacer match con data 1`] = `
<div>
  <h1>Quien es el Pokemon?</h1><!-- Imagen -->
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub><!-- Opciones -->
  <!--selection-pokemon-->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
